### PR TITLE
Adding e2e test for Nodeport on Windows

### PR DESCRIPTION
**Pull request by immuzz**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

/sig windows
/sig testing

Contributor : @brusmx

**What this PR does / why we need it**: 
Tests nodeport for Windows

**Which issue(s) this PR fixes**:  

Fixes # https://github.com/kubernetes/kubernetes/issues/73327

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by immuzz** [comment]
This is test is taking ~3 minutes to pass!

**Comment by immuzz** [comment]
@yujuhong @dineshgovindasamy @PatrickLang Please review

**Comment by michmike** [comment]
cc:  @neolit123  for review as well

File: test/e2e/windows/service.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (neolit123) [inline test/e2e/windows/service.go]: 2019

File: test/e2e/windows/service.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+			svc.Spec.Type = v1.ServiceTypeNodePort
+		})
+		jig.SanityCheckService(service, v1.ServiceTypeNodePort)
+		nodePort := int(service.Spec.Ports[0].NodePort)
+
+		By("creating pod to be part of service " + serviceName)
+		jig.RunOrFail(ns, nil)
+		
+		// Create a testing pod so we can ping the nodeport from within the cluster
+		By(fmt.Sprintf("creating testing pod to curl http://%s:%d", nodeIP, nodePort))
```
Comment (neolit123) [inline test/e2e/windows/service.go]: `Pod` uppercase.

File: test/e2e/windows/service.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/pkg/securitycontext"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	defaultCmd          = "curl.exe -s -o 127.0.0.1"
```
Comment (neolit123) [inline test/e2e/windows/service.go]: this command is not valid.

`curl.exe` will work as long as it is in `PATH`. is it?

`-s` means silent and this flag is OK
`-o` means write output to a stream or file.

this is valid:
`curl -s 127.0.0.1`

but i'm not sure about the intent of the default command.

File: test/e2e/windows/service.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+var _ = SIGDescribe("Services", func() {
+	f := framework.NewDefaultFramework("services")
+
+	var cs clientset.Interface
+
+	BeforeEach(func() {
+		cs = f.ClientSet
+	})
+	//Note(immuzz -> Muzz Imam ) This test will run only for Windows Containers
```
Comment (neolit123) [inline test/e2e/windows/service.go]: i would change this whole comment to:
> // Windows containers only

File: test/e2e/windows/service.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+		nodeIP := framework.PickNodeIP(jig.Client) // for later
+
+		By("creating service " + serviceName + " with type=NodePort in namespace " + ns)
+		service := jig.CreateTCPServiceOrFail(ns, func(svc *v1.Service) {
+			svc.Spec.Type = v1.ServiceTypeNodePort
+		})
+		jig.SanityCheckService(service, v1.ServiceTypeNodePort)
+		nodePort := int(service.Spec.Ports[0].NodePort)
+
+		By("creating pod to be part of service " + serviceName)
```
Comment (neolit123) [inline test/e2e/windows/service.go]: `Pod`

File: test/e2e/windows/service.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+		jig.RunOrFail(ns, nil)
+		
+		// Create a testing pod so we can ping the nodeport from within the cluster
+		By(fmt.Sprintf("creating testing pod to curl http://%s:%d", nodeIP, nodePort))
+		windowsSelector := map[string]string{"beta.kubernetes.io/os": "windows"}
+		winPodSpec := defaultWindowsPod.DeepCopy()
+		winPodSpec.Spec.NodeSelector = windowsSelector
+		curl := fmt.Sprintf("curl.exe -s -o /dev/null -w \"%%{http_code}\" http://%s:%d", nodeIP, nodePort)
```
Comment (neolit123) [inline test/e2e/windows/service.go]: there is no `/dev/null` on Windows, but there is `NUL`

i'm going to assume we are trying to get the HTTP status?
try:
```
curl := fmt.Sprintf(`curl.exe -s -o NUL -w "%{http_code}" http://%s:%d`, nodeIP, nodePort)
```

but one question - if we are duplicating the Pod and always overwriting the default command, why do we need the default command?

File: test/e2e/windows/service.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+		winPodSpec.Spec.Containers[0].Args = []string{"cmd", "/c", curl}
+		pod, err := cs.CoreV1().Pods(ns).Create(winPodSpec)
+		Expect(err).NotTo(HaveOccurred())
+		
+		By("waiting for pod to be running")
+		err = f.WaitForPodRunning(pod.Name)
+		Expect(err).NotTo(HaveOccurred(),
+			"Error waiting for pod %s to run", pod.Name)
```
Comment (neolit123) [inline test/e2e/windows/service.go]: `Pod`

File: test/e2e/windows/service.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+		
+		By("obtaining the logs of the command")
+		logs, err := framework.GetPodLogs(cs, ns, pod.Name, pod.Spec.Containers[0].Name)
+		Expect(err).NotTo(HaveOccurred(),
+			"Error getting logs from pod %s in namespace %s", pod.Name, ns)
```
Comment (neolit123) [inline test/e2e/windows/service.go]: `Pod`

**Review by neolit123 — COMMENTED** [review]
thanks for the PR @immuzz 
in general the test looks OK to me.

mostly added some minor comments about curl usage.

File: test/e2e/windows/service.go (hunk 6 above)
Comment (brusMX) [reply test/e2e/windows/service.go]: At the beginning we were using the default command just to put a sleep and be able to exist. You are right, if this is not used in other context there is no need to set it up. Removing it.

Comment (brusMX) [reply test/e2e/windows/service.go]: Also:
 
```
-o /dev/null
```

And 

```
-o NUL
```

both behave the same way on both `powershell` and in `cmd`. But if you want we can adopt your way.

**Review by michmike — APPROVED** [review]

File: test/e2e/windows/service.go (hunk 6 above)
Comment (neolit123) [reply test/e2e/windows/service.go]: doing `curl.exe -s -o /dev/null` will create the actual path e.g. `C:\dev\null`.

sadly, powershell and cmd use different variables for the null device:
PS uses `$null`
cmd uses `NUL`.

and one does not work for the other.
if this test is supposed to work for both shells, just pipe to a local file (but not /dev/null):

`curl.exe s -o ./curl-output`

/hold

Comment (michmike) [reply test/e2e/windows/service.go]: @neolit123  i am not seeing the same behavior. on my windows 10 laptop. when i run `curl -s -o /dev/null -w "%{http_code}" "https://img-s-msn-com.akamaized.net/tenant/amp/entityid/BBUmCRP.img?h=1224&w=799&m=6&q=60&o=f&l=f&x=1064&y=935"`, even with admin privileges the c:/dev/null folder is not created. the output is on the console

Comment (brusMX) [reply test/e2e/windows/service.go]: Sorry guys the changes are done, but seems like @immuzz didnt give me contributor access to his repo. Also remember this is 1809 windows, which does behave differently.
Waiting for the access you can review the code here:
https://github.com/kubernetes/kubernetes/compare/master...immuzz:origin/nodeportwindows

Comment (neolit123) [reply test/e2e/windows/service.go]: it's created for me on both Windows 7 and Windows 10.
but there is definitely no `/dev/null` on Windows:
https://stackoverflow.com/questions/313111/is-there-a-dev-null-on-windows

to be on the safe side and given we are using `cmd -c` we should just do what i recommended in the first comment:
```
curl.exe -s -o NUL -w "%{http_code}" http://%s:%d`, nodeIP, nodePort
```

Comment (neolit123) [reply test/e2e/windows/service.go]: @brusMX if you'd like just close this PR using `/close` and open a new one with the changes.
but please squash the commits into a single one.

Comment (neolit123) [reply test/e2e/windows/service.go]: > Also remember this is 1809 windows, which does behave differently.

i have 1809 here too:
null-device-windows

**Comment by neolit123** [comment]
> added cncf-cla: no  and removed cncf-cla: yes  labels 2 minutes ago

looks like the rebase picked some extra commits.

**Comment by brusMX** [comment]
Moving to https://github.com/kubernetes/kubernetes/pull/74938

**Comment by neolit123** [comment]
/close
please ping me on the new one. thanks.

**Pull request closed without merging**